### PR TITLE
lib.converge: optimise

### PR DESCRIPTION
**Pull request by alyssais**
###### Motivation for this change

https://github.com/NixOS/nixpkgs/pull/51884#pullrequestreview-223602570

Cc: @cstrahan

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by cstrahan** [comment]
Awesome -- thanks!

**Comment by edolstra** [comment]
What's the use case for `converge`? Iterating until convergence sounds like the sort of thing we should *not* be doing in Nixpkgs (it's slow and memory-hungry enough as is).

**Comment by alyssais** [comment]
> What's the use case for `converge`?

https://github.com/NixOS/nixpkgs/commit/67b1265fb3d38ead5a57fee838405a2d997777c2#diff-10fc6e70aec47edd191208687e4839b2

tl;dr, it's used for recursively resolving dependencies. It's a primitive that things like closePropagation could be implemented on top of.
